src/services/jsTyping.ts

```diff
@@ -0,0 +1,271 @@
+            const packageJsonFiles =
+                host.readDirectory(nodeModulesPath, /*extension*/ undefined, /*exclude*/ undefined, /*depth*/ 2).filter(f => ts.getBaseFileName(f) === "package.json");
+            for (const packageJsonFile of packageJsonFiles) {
+                const packageJsonDict = tryParseJson(packageJsonFile, host);
+                if (!packageJsonDict) { continue; }
+
+                filesToWatch.push(packageJsonFile);
+
+                // npm 3 has the package.json contains a "_requiredBy" field
+                // we should include all the top level module names for npm 2, and only module names whose
+                // "_requiredBy" field starts with "#" or equals "/" for npm 3.
+                if (packageJsonDict._requiredBy &&
+                    packageJsonDict._requiredBy.filter((r: string) => r[0] === "#" || r === "/").length === 0) {
+                    continue;
+                }
+
+                // If the package has its own d.ts typings, those will take over. Otherwise the package name will be used
+                // to download d.ts files from DefinitelyTyped
+                const packageName = packageJsonDict["name"];
+                if (packageJsonDict.hasOwnProperty("typings")) {
+                    const absPath = ts.getNormalizedAbsolutePath(packageJsonDict.typings, ts.getDirectoryPath(packageJsonFile));
+                    inferredTypings[packageName] = absPath;
+                }
+                else {
+                    typingNames.push(packageName);
+                }
+            }
+            mergeTypings(typingNames);
+        }
+
+        function getTypingNamesFromCompilerOptions(options: CompilerOptions) {
+            const typingNames: string[] = [];
+            if (!options) {
+                return;
+            }
+
+            if (options.jsx === JsxEmit.React) {
+                typingNames.push("react");
+            }
+            if (options.moduleResolution === ModuleResolutionKind.NodeJs) {
+                typingNames.push("node");
+            }
+            mergeTypings(typingNames);
+        }
+    }
+
+    /**
+     * Keep a list of typings names that we knew cannot be obtained at the moment (could be because 
+     * of network issues or because the package doesn't hava a d.ts files in DefinitelyTyped), so 
+     * that we won't try again next time within thie session.
+     * @param cachedTypingsPaths The list of resolved cached d.ts paths
+     * @param newTypings The list of new typings that the host attempted to acquire using TSD
+     * @param cachePath The path to the local tsd.json cache
+     */
+    export function updateNotFoundTypingNames(triedTypingNames: string[], cachePath: string, host: HostType): void {
+        const tsdJsonPath = ts.combinePaths(cachePath, "tsd.json");
+        const cacheTsdJsonDict = tryParseJson(tsdJsonPath, host);
+        if (cacheTsdJsonDict) {
+            if (cacheTsdJsonDict.hasOwnProperty("installed")) {

```

Comment:
We always need to check for missing typings even when there aren't any installed. (I hit this case for the Cordova project).  Here is how I fixed it locally:

 let installedTypingFiles: string[] = [];
        if (cacheTsdJsonDict) {
            if (cacheTsdJsonDict.hasOwnProperty("installed")) {
                installedTypingFiles = Object.keys(cacheTsdJsonDict.installed);
            }